src/Test/Utilities/Shared/Compilation/IRuntimeUtility.cs

```diff
@@ -0,0 +1,253 @@
+            {
+                EmitReferences(current, fullNameSet, dependencies, diagnostics);
+            }
+
+            return RuntimeUtilities.EmitCompilationCore(compilation, manifestResources, diagnostics, testData);
+        }
+
+        internal static EmitOutput? EmitCompilationCore(
+            Compilation compilation,
+            IEnumerable<ResourceDescription> manifestResources,
+            DiagnosticBag diagnostics,
+            CompilationTestData testData
+        )
+        {
+            using (var executableStream = new MemoryStream())
+            {
+                var pdb = default(ImmutableArray<byte>);
+                var assembly = default(ImmutableArray<byte>);
+                var pdbStream = MonoHelpers.IsRunningOnMono()
+                    ? null
+                    : new MemoryStream();
+
+                EmitResult result;
+                try
+                {
+                    result = compilation.Emit(
+                        executableStream,
+                        pdbStream: pdbStream,
+                        xmlDocumentationStream: null,
+                        win32Resources: null,
+                        manifestResources: manifestResources,
+                        options: EmitOptions.Default,
+                        debugEntryPoint: null,
+                        testData: testData,
+                        getHostDiagnostics: null,
+                        cancellationToken: default(CancellationToken));
+                }
+                finally
+                {
+                    if (pdbStream != null)
+                    {
+                        pdb = pdbStream.ToImmutable();
+                        pdbStream.Dispose();
+                    }
+                }
+
+                diagnostics.AddRange(result.Diagnostics);
+                assembly = executableStream.ToImmutable();
+
+                if (result.Success)
+                {
+                    return new EmitOutput(assembly, pdb);
+                }
+
+                return null;
+            }
+        }
+    }
+
+    public interface IRuntimeUtility: System.IDisposable
+    {
+        void Emit(Microsoft.CodeAnalysis.Compilation mainCompilation, IEnumerable<ResourceDescription> manifestResources, bool usePdbForDebugging = false);
+        int Execute(string moduleName, string expectedOutput);
+        ImmutableArray<byte> GetMainImage();
+        ImmutableArray<byte> GetMainPdb();
+        ImmutableArray<Diagnostic> GetDiagnostics();
+        SortedSet<string> GetMemberSignaturesFromMetadata(string fullyQualifiedTypeName, string memberName);
+        IList<ModuleData> GetAllModuleData();
+        void PeVerify();
+        string[] PeVerifyModules(string[] modulesToVerify, bool throwOnError = true);
+    }
+
+    internal interface IInternalRuntimeUtility
+    {
+        CompilationTestData GetCompilationTestData();
+    }
```

Comment:
The implementations are named `...RuntimeEnvironment`. Consider renaming the interfaces `IRuntimeEnvironment` and `IInternalRuntimeEnvironment`.